src/mail_data.c

```diff
@@ -0,0 +1,182 @@
+                gSaveBlock1Ptr->mail[id].words[i] = 0xFFFF;
+            for (i = 0; i < PLAYER_NAME_LENGTH - 1 && gSaveBlock2Ptr->playerName[i] != EOS; i++)
+                gSaveBlock1Ptr->mail[id].playerName[i] = gSaveBlock2Ptr->playerName[i];
+            for (; i <= 5; i++)
+            	gSaveBlock1Ptr->mail[id].playerName[i] = CHAR_SPACE;
+            gSaveBlock1Ptr->mail[id].playerName[i] = EOS;
+            for (i = 0; i < 4; i++)
+                gSaveBlock1Ptr->mail[id].trainerId[i] = gSaveBlock2Ptr->playerTrainerId[i];
+            species = GetBoxMonData(&mon->box, MON_DATA_SPECIES);
+            personality = GetBoxMonData(&mon->box, MON_DATA_PERSONALITY);
+            gSaveBlock1Ptr->mail[id].species = SpeciesToMailSpecies(species, personality);
+            gSaveBlock1Ptr->mail[id].itemId = itemId;
+            SetMonData(mon, MON_DATA_MAIL, &id);
+            SetMonData(mon, MON_DATA_HELD_ITEM, heldItem);
+            return id;
+        }
+    }
+    return 0xFF;
+}
+
+u16 SpeciesToMailSpecies(u16 species, u32 personality)
+{
+    return species == 0xC9 ? ((GetUnownLetterByPersonality(personality) << 16) + 0x75300000) >> 16 : species;
+}
+
+u16 MailSpeciesToSpecies(u16 mailSpecies, u16 *buffer)
+{
+    u16 result;
+
+    if (mailSpecies >= 30000 && mailSpecies < (30000 + UNOWN_FORM_COUNT))
+    {
+        result = SPECIES_UNOWN;
+        *buffer = mailSpecies - 30000;
+    }
+    else
+    {
+        result = mailSpecies;
+    }
+    return result;
+}
+
+u8 GiveMailToMon2(struct Pokemon *mon, struct MailStruct *mail)
+{
+    u8 heldItem[2];
+    u16 itemId = mail->itemId;
+    u8 mailId = GiveMailToMon(mon, itemId);
+
+    if (mailId == 0xFF)
+        return 0xFF;
+    gSaveBlock1Ptr->mail[mailId] = *mail;
+    SetMonData(mon, MON_DATA_MAIL, &mailId);
+    heldItem[0] = itemId;
+    heldItem[1] = itemId >> 8;
+
+    SetMonData(mon, MON_DATA_HELD_ITEM, heldItem);
+
+    return mailId;
+}
+
+static bool32 DummyMailFunc(void)
+{
+    return FALSE;
+}
+
+void TakeMailFromMon(struct Pokemon *mon)
+{
+    u8 heldItem[2];
+    u8 mailId;
+
+    if (MonHasMail(mon))
+    {
+        mailId = GetMonData(mon, MON_DATA_MAIL);
+        gSaveBlock1Ptr->mail[mailId].itemId = ITEM_NONE;
+        mailId = 0xFF;
+        heldItem[0] = ITEM_NONE;
+        heldItem[1] = ITEM_NONE << 8;
+        SetMonData(mon, MON_DATA_MAIL, &mailId);
+        SetMonData(mon, MON_DATA_HELD_ITEM, heldItem);
+    }
+}
+
+void ClearMailItemId(u8 mailId)
+{
+    gSaveBlock1Ptr->mail[mailId].itemId = ITEM_NONE;
+}
+
+u8 TakeMailFromMon2(struct Pokemon *mon)
+{
+    u8 i, newMailId;
+    u8 newHeldItem[2];
+
+    newHeldItem[0] = ITEM_NONE;
+    newHeldItem[1] = ITEM_NONE << 8;
+    newMailId = 0xFF;
+    for (i = PARTY_SIZE; i < MAIL_COUNT; i++)
+    {
+        if (gSaveBlock1Ptr->mail[i].itemId == ITEM_NONE)
+        {
+            memcpy(&gSaveBlock1Ptr->mail[i], &gSaveBlock1Ptr->mail[GetMonData(mon, MON_DATA_MAIL)], sizeof(struct MailStruct));
```

Comment:
If you do `gSaveBlock1Ptr->mail[i] = gSaveBlockPtr->mail[GetMonData(mon, MON_DATA_MAIL)];`, does it still match?

Reply:
No. I corrected `gSaveBlockPtr` to `gSaveBlock1Ptr` and it still doesn't match